vyper/context/validation/local.py

```diff
@@ -112,15 +116,24 @@ def __init__(
         self.func = namespace["self"].get_member(fn_node.name, fn_node)
         namespace.update(self.func.arguments)
 
-        if not self.func.is_public:
+        if self.func.visibility is FunctionVisibility.PRIVATE:
             node_list = fn_node.get_descendants(
                 vy_ast.Attribute, {"value.id": "msg", "attr": "sender"}
             )
             if node_list:
                 raise ConstancyViolation(
                     "msg.sender is not allowed in private functions", node_list[0]
                 )
-        if not getattr(self.func, "is_payable", False):
+        if self.func.mutability is StateMutability.PURE:
+            node_list = fn_node.get_descendants(
+                vy_ast.Attribute, {"value.id": {"self", "msg", "tx", "block", "chain"}}
+            )
+            if node_list:
+                raise ConstancyViolation(
+                    "not allowed to query contract or environment variables in pure functions",
+                    node_list[0],
+                )
```

Comment:
This check is new for the `pure` decorator

Reply:
I like it.  I'm not sure if it's actually better, but maybe we could pull that list of objects from `contract/environment.py` so that if the environment variables change in the future, we don't have to update this line as well?

Reply:
`self` won't be anywhere in `environment.py` because it's special and magic, but I think the others would be available somewhere there.

Reply:
Yeah, I was thinking that. Does one exist? or just get the keys of something?